src/resolve-config.js

```diff
@@ -1,31 +1,76 @@
 "use strict";
 
 const cosmiconfig = require("cosmiconfig");
+const editorconfig = require("editorconfig");
+const mem = require("mem");
 const minimatch = require("minimatch");
 
 const asyncWithCache = cosmiconfig("prettier");
 const asyncNoCache = cosmiconfig("prettier", { cache: false });
 const syncWithCache = cosmiconfig("prettier", { sync: true });
 const syncNoCache = cosmiconfig("prettier", { cache: false, sync: true });
 
+const editorconfigAsyncNoCache = filePath =>
```

Comment:
This is starting to blow up combinatorially. Any way we can apply some polymorphism magic to clean it up a bit?

Reply:
🤔

```js
const asyncWithCache = configgers.cosmiconfig.async.cache;
const asyncNoCache = configgers.cosmiconfig.async.noCache;
const syncWithCache = configgers.cosmiconfig.sync.cache;
const syncNoCache = configgers.cosmiconfig.sync.noCache;
const editorconfigAsyncWithCache = configgers.editorconfig.async.cache;
const editorconfigAsyncNoCache = configgers.editorconfig.async.noCache;
const editorconfigSyncWithCache = configgers.editorconfig.sync.cache;
const editorconfigSyncNoCache = configgers.editorconfig.sync.noCache;
```
```js
const configgers = createConfiggers([
  {
    cosmiconfig: { type: "cosmiconfig" },
    editorconfig: { type: "editorconfig" }
  },
  {
    sync: { sync: true },
    async: { sync: false }
  },
  {
    cache: { cache: true },
    noCache: { cache: false }
  }
]);

function createConfiggers(configgerArgs) {
  const createdConfiggers = {};

  recursive(null, createdConfiggers, configgerArgs, []);

  return createdConfiggers;

  function recursive(parent, target, args, stack) {
    if (args.length !== 0) {
      Object.keys(args[0]).forEach(key => {
        const option = args[0][key];
        const subTarget = (target[key] = {});
        recursive(
          target,
          subTarget,
          args.slice(1),
          stack.concat({ key, option })
        );
      });
    } else {
      const options = stack.reduce(
        (current, element) => Object.assign(current, element.option),
        {}
      );
      const lastKey = stack[stack.length - 1].key;
      parent[lastKey] = createConfigger(
        options.type,
        options.sync,
        options.cache
      );
    }
  }
}

function createConfigger(type, sync, cache) {
  if (type === "cosmiconfig") {
    return cosmiconfig("prettier", { sync, cache });
  }
  const editorConfigger = !sync
    ? filePath =>
        filePath && editorconfig.parse(filePath).then(editorConfigToPrettier)
    : filePath =>
        filePath && editorConfigToPrettier(editorconfig.parseSync(filePath));
  return cache ? mem(editorConfigger) : editorConfigger;
}
```

Reply:
@ikatyang I think the "manual" way is better in this case :)